### ISSUE TITLE
Vapor 3's Client does not handle redirects

### ISSUE THREAD

**Issue by gwynne** [vapor/vapor#1461]
As most of us know, the `3xx` series of HTTP response codes specifies redirection. It is common for HTTP libraries to support transparent and/or configurable handling of redirects without the need for the application layer to understand the nuances of handing the `Location` response header and the semantics of the various 3xx responses. In `URLSession` this is handled by making a delegate callback each time a redirect takes place, with redirects chaining indefinitely by default (this is not a good default configuration; there should be a default limit to prevent DoS due to redirect loops).

In short, Vapor 3's Client should support configurable redirect handling. Of particular interest is the ability to retrieve the final URL at the end of a redirect chain (such as `libcurl` allows via `CURLINFO_EFFECTIVE_URL`).

Notably, Vapor 2's `EngineClient` had no redirect support at all, while the `FoundationClient` inherited `URLSession`'s default behavior.

**Comment by Joannis** [comment]
See #1480
